Avoid recursive jump for better loop detection

qwe661234 opened the pull request:
Previously, we employed recursive jump translation to implement an extended basic block. Nevertheless, this approach makes it challenging to detect loop paths since we position the loop's entry block inside the block, and its using frequency would not be updated.

Based on this observation, it becomes necessary to eliminate the recursive jump translation. By doing so, we can accurately update the using frequency of the loop's entry block.

As shown in the performance results below, we gain 4% performance improvement when running coreMark and 37% when running SciMark2, but we lost 1% performance when running dhrysone.

* Intel Core i7-11700

|   Metric   | origin  | proposed |Speedup|
|--------- |--------- |--------- |--------- |
| CoreMark   | 2193.28 | 2289.26  |  +4%  |
| SciMark2   |   13.45 |   18.48  |  +37%  |
| Dhrystone  | 1413.11 | 1447.11  |  -1%  |

See: #142

jserv commented on src/emulate.c:
```diff
@@ -405,19 +405,8 @@ static void block_translate(riscv_t *rv, block_t *block)
-                ir->branch_taken = ir + 1;
-                continue;
-            }
+        if (insn_is_branch(ir->opcode))
```
Shall we change `master` branch accordingly?

qwe661234 replied:
Yes, this change is also required for detecting memset and memcpy.

jserv commented:
> Metric	origin	proposed	Speedup
> SciMark2	13.45	18.48	+4%

Can you check the numbers of SciMark2?
(18.48 - 13.45) / 13.45 = 37.39%

Commit:
Avoid recursive jump for better loop detection

Previously, we employed recursive jump translation to implement an extended
basic block. Nevertheless, this approach makes it challenging to detect
loop paths since we position the loop's entry block inside the block, and
its using frequency would not be updated.

Based on this observation, it becomes necessary to eliminate the recursive
jump translation. By doing so, we can accurately update the using frequency
of the loop's entry block.

As shown in the performance results below, we gain 4% performance improvement
when running coreMark and 37% when running SciMark2, but we lost 1% performance
when running dhrysone.

* Intel Core i7-11700

| Metric     | origin  | proposed |Speedup|
|------------+---------+----------+-------|
| CoreMark   | 2193.28 |  2289.26 |   +4% |
| SciMark2   |   13.45 |    18.48 |  +37% |
| Dhrystone  | 1413.11 |  1447.11 |   -1% |

See: #142

qwe661234 commented:
> > Metric	origin	proposed	Speedup
> > SciMark2	13.45	18.48	+4%
> 
> Can you check the numbers of SciMark2? (18.48 - 13.45) / 13.45 = 37.39%

I Update this in git log.
